artiq/coredevice/mirny.py

```diff
@@ -140,3 +149,160 @@ def write_ext(self, addr, length, data):
+        :param output_on - RF output on (bool)
+        """
+        if not 4 >= channel >= 1:
+            raise ValueError("channel must be 1, 2, 3 or 4")
+        if not 63 >= att_mu >= 0:
+            raise ValueError("Invalid Almazny attenuator setting")
+
+        att_ins = [ (att_mu >> i) & 0x01 for i in range(6) ]
+        shift_reg = [0, 1 if output_on else 0] + att_ins
+
+        self._send_mezz_data([
+            (ALMAZNY_REG_LATCH_BASE + (channel - 1), 0)
+        ])
+        for val in shift_reg:
+            self._cycle([
+                (ALMAZNY_SER_MOSI, val) 
+            ])
+        self._latch(channel)
+
+    @kernel
+    def reg_clear(self, channel):
+        """
+        Clears content of a register.
+        :param channel - index of the register [1-4]
+        """
+        if not 4 >= channel >= 1:
+            raise ValueError("register must be 1, 2, 3 or 4")
+        self._send_mezz_data([
+            (ALMAZNY_REG_CLEAR, 0),
+            (ALMAZNY_REG_LATCH_BASE + (channel - 1), 0)
+        ])
+        self._send_mezz_data([
+            (ALMAZNY_REG_CLEAR, 0),
+            (ALMAZNY_REG_LATCH_BASE + (channel - 1), 1)
+        ])
+        self._send_mezz_data([
+            (ALMAZNY_REG_CLEAR, 1),
+            (ALMAZNY_REG_LATCH_BASE + (channel - 1), 0)
+        ])
+
+    @kernel
+    def reg_clear_all(self):
+        """
+        Clears all registers.
+        """
+        self._send_mezz_data([
+            (ALMAZNY_REG_CLEAR, 0),
+            (ALMAZNY_REG_LATCH_BASE, 0),
+            (ALMAZNY_REG_LATCH_BASE + 1, 0),
+            (ALMAZNY_REG_LATCH_BASE + 2, 0),
+            (ALMAZNY_REG_LATCH_BASE + 3, 0),
+        ])
+        self._send_mezz_data([
+            (ALMAZNY_REG_CLEAR, 0),
+            (ALMAZNY_REG_LATCH_BASE, 1),
+            (ALMAZNY_REG_LATCH_BASE + 1, 1),
+            (ALMAZNY_REG_LATCH_BASE + 2, 1),
+            (ALMAZNY_REG_LATCH_BASE + 3, 1),
+        ])
+        self._send_mezz_data([
+            (ALMAZNY_REG_CLEAR, 1),
+            (ALMAZNY_REG_LATCH_BASE, 0),
+            (ALMAZNY_REG_LATCH_BASE + 1, 0),
+            (ALMAZNY_REG_LATCH_BASE + 2, 0),
+            (ALMAZNY_REG_LATCH_BASE + 3, 0),
+        ])
+
+    @kernel
+    def _cycle(self, data):
+        """
+        one cycle for inputting register data
+        """
+        self._send_mezz_data([
+            (ALMAZNY_SER_CLK, 0),
+        ] + data)
+        self._send_mezz_data([
+            (ALMAZNY_SER_CLK, 1),
+        ] + data)
+
+    @kernel
+    def _latch(self, channel):
+        self._send_mezz_data([
+            (ALMAZNY_SER_CLK, 1),
+            (ALMAZNY_REG_LATCH_BASE + (channel - 1), 1)
+        ])
+        self._send_mezz_data([
+            (ALMAZNY_REG_LATCH_BASE + (channel - 1), 0)
+        ])
+
+    @kernel
+    def put_data(self, pin, d):
+        self.mezz_data = (self.mezz_data & ~(1 << pin)) | d << pin  # data
+        self.mezz_data |= 1 << pin + 8  # oe
```

Comment:
Avoid those side effects. Just build an integer locally and pass it around.

Reply:
And I suggest having a constant OE mask, possibly defined in the module which is systematically ORed.

Reply:
If you want to retain state, then retain high-level state e.g. ``self.rf_switches = [False]*4; self.att_mu = [0x3f]*4``. Then at each update build the control word based on the high level state, and send it to the hardware.